File: src/class013/C_QueueStackAndCircularQueue.cpp
#include <queue>
#include <stack>
#include <vector>

// 使用 C++ 的内部实现，内部通常使用双向链表
class Queue1 {
public:
    std::queue<int> queue;

    bool isEmpty() {
        return queue.empty();
    }

    void offer(int num) {
        queue.push(num);
    }

    int poll() {
        int front = queue.front();
        queue.pop();
        return front;
    }

    int peek() {
        return queue.front();
    }

    int size() {
        return queue.size();
    }
};

// 使用静态数组来实现队列
class Queue2 {
private:
    std::vector<int> queue;
    int l, r, capacity;

public:
    Queue2(int n) : queue(std::vector<int>(n)), l(0), r(0), capacity(n) {}

    bool isEmpty() {
        return l == r;
    }

    void offer(int num) {
        queue[r] = num;
        r = (r + 1) % capacity;
    }

    int poll() {
        int val = queue[l];
        l = (l + 1) % capacity;
        return val;
    }

    int head() {
        return queue[l];
    }

    int tail() {
        return queue[(r - 1 + capacity) % capacity];
    }

    int size() {
        return (r - l + capacity) % capacity;
    }
};

// 使用 C++ 内部的 stack 实现
class Stack1 {
private:
    std::stack<int> stack;

public:
    bool isEmpty() {
        return stack.empty();
    }

    void push(int num) {
        stack.push(num);
    }

    int pop() {
        int top = stack.top();
        stack.pop();
        return top;
    }

    int peek() {
        return stack.top();
    }

    int size() {
        return stack.size();
    }
};

// 使用静态数组来实现栈
class Stack2 {
private:
    std::vector<int> stack;
    int top;

public:
    Stack2(int n) : stack(std::vector<int>(n)), top(0) {}

    bool isEmpty() {
        return top == 0;
    }

    void push(int num) {
        stack[top++] = num;
    }

    int pop() {
        return stack[--top];
    }

    int peek() {
        return stack[top - 1];
    }

    int size() {
        return top;
    }
};

// 实现循环队列
class MyCircularQueue {
private:
    std::vector<int> queue;
    int l, r, size, limit;

public:
    MyCircularQueue(int k) : queue(std::vector<int>(k)), l(0), r(0), size(0), limit(k) {}

    bool enQueue(int value) {
        if (isFull()) {
            return false;
        } else {
            queue[r] = value;
            r = (r + 1) % limit;
            size++;
            return true;
        }
    }

    bool deQueue() {
        if (isEmpty()) {
            return false;
        } else {
            l = (l + 1) % limit;
            size--;
            return true;
        }
    }

    int Front() {
        return isEmpty() ? -1 : queue[l];
    }

    int Rear() {
        if (isEmpty()) {
            return -1;
        } else {
            int last = r == 0 ? limit - 1 : r - 1;
            return queue[last];
        }
    }

    bool isEmpty() {
        return size == 0;
    }

    bool isFull() {
        return size == limit;
    }
};

File: src/class013/C_QueueStackAndCircularQueue.py
from collections import deque

# 使用 Python 内置的双端队列（deque）来实现队列
class Queue1:
    def __init__(self):
        self.queue = deque()

    def is_empty(self):
        return not self.queue

    def offer(self, num):
        self.queue.append(num)

    def poll(self):
        return self.queue.popleft()

    def peek(self):
        return self.queue[0]

    def size(self):
        return len(self.queue)

# 使用静态数组来实现队列
class Queue2:
    def __init__(self, n):
        self.queue = [0] * n
        self.l = 0
        self.r = 0
        self.capacity = n

    def is_empty(self):
        return self.l == self.r

    def offer(self, num):
        self.queue[self.r] = num
        self.r = (self.r + 1) % self.capacity

    def poll(self):
        val = self.queue[self.l]
        self.l = (self.l + 1) % self.capacity
        return val

    def head(self):
        return self.queue[self.l]

    def tail(self):
        return self.queue[(self.r - 1 + self.capacity) % self.capacity]

    def size(self):
        return (self.r - self.l + self.capacity) % self.capacity

# 使用 Python 的内置栈（实际上是使用 list 实现）
class Stack1:
    def __init__(self):
        self.stack = []

    def is_empty(self):
        return not self.stack

    def push(self, num):
        self.stack.append(num)

    def pop(self):
        return self.stack.pop()

    def peek(self):
        return self.stack[-1]

    def size(self):
        return len(self.stack)

# 使用静态数组来实现栈
class Stack2:
    def __init__(self, n):
        self.stack = [0] * n
        self.top = 0

    def is_empty(self):
        return self.top == 0

    def push(self, num):
        self.stack[self.top] = num
        self.top += 1

    def pop(self):
        self.top -= 1
        return self.stack[self.top]

    def peek(self):
        return self.stack[self.top - 1]

    def size(self):
        return self.top

# 实现循环队列
class MyCircularQueue:
    def __init__(self, k):
        self.queue = [0] * k
        self.l = 0
        self.r = 0
        self.size = 0
        self.limit = k

    def en_queue(self, value):
        if self.is_full():
            return False
        else:
            self.queue[self.r] = value
            self.r = (self.r + 1) % self.limit
            self.size += 1
            return True

    def de_queue(self):
        if self.is_empty():
            return False
        else:
            self.l = (self.l + 1) % self.limit
            self.size -= 1
            return True

    def front(self):
        if self.is_empty():
            return -1
        else:
            return self.queue[self.l]

    def rear(self):
        if self.is_empty():
            return -1
        else:
            last = 0 if self.r == 0 else self.r - 1
            return self.queue[last]

    def is_empty(self):
        return self.size == 0

    def is_full(self):
        return self.size == self.limit
